animations-dbus/animations-dbus-server-surface.c

```diff
@@ -0,0 +1,691 @@
+  g_autoptr(GVariant) available_effects =
+    g_variant_ref_sink (animations_dbus_server_surface_bridge_get_available_effects (priv->bridge));
+
+  animations_dbus_animatable_surface_complete_list_effects (animatable_surface,
+                                                            invocation,
+                                                            available_effects);
+  return TRUE;
+}
+
+static void
+animations_dbus_animatable_surface_interface_init (AnimationsDbusAnimatableSurfaceIface *iface)
+{
+  iface->handle_attach_animation_effect = animations_dbus_server_surface_attach_animation_effect;
+  iface->handle_detach_animation_effect = animations_dbus_server_surface_detach_animation_effect;
+  iface->handle_list_effects = animations_dbus_server_surface_list_animation_effects;
+}
+
+static void
+animations_dbus_server_surface_set_property (GObject      *object,
+                                             guint         prop_id,
+                                             const GValue *value,
+                                             GParamSpec   *pspec)
+{
+  AnimationsDbusServerSurface *server_surface = ANIMATIONS_DBUS_SERVER_SURFACE (object);
+  AnimationsDbusServerSurfacePrivate *priv = animations_dbus_server_surface_get_instance_private (server_surface);
+
+  switch (prop_id)
+    {
+    case PROP_CONNECTION:
+      priv->connection = g_value_dup_object (value);
+      break;
+    case PROP_SERVER:
+      /* Observing reference, AnimationsDbusServer is expected to last
+       * longer than AnimationsDbusServerSurface */
+      priv->server = g_value_get_object (value);
+      break;
+    case PROP_BRIDGE:
+      priv->bridge = g_value_dup_object (value);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+    }
+}
+
+static GVariant *
+serialize_attached_effects_to_variant (GHashTable *effects_for_events)
+{
+  g_auto(GVariantDict) vardict;
+  gpointer key, value;
+  GHashTableIter iter;
+
+  g_variant_dict_init (&vardict, NULL);
+
+  g_hash_table_iter_init (&iter, effects_for_events);
+  while (g_hash_table_iter_next (&iter, &key, &value))
+    {
+      const char *event = key;
+      GQueue *effects = value;
+      g_autoptr(GPtrArray) names = g_ptr_array_new_full (g_queue_get_length (effects), NULL);
+
+      for (GList *link = g_queue_peek_head_link (effects);
+           link != NULL;
+           link = link->next)
+        {
+          AttachedEffectInfo *info = link->data;
+          g_ptr_array_add (names,
+                           (gpointer) g_dbus_interface_skeleton_get_object_path (G_DBUS_INTERFACE_SKELETON (info->server_effect)));
+        }
+
+      g_variant_dict_insert_value (&vardict,
+                                   event,
+                                   g_variant_new_strv ((const char * const *) names->pdata,
+                                                       names->len));
```

Comment:
Yeah, using a builder here also works.

Reply:
Hmm, something strange has started happening since I switched to using GVariantBuilder here:

```
(gnome-shell:65911): GLib-CRITICAL **: g_variant_builder_add_value: assertion '!GVSB(builder)->expected_type || g_variant_is_of_type (value, GVSB(builder)->expected_type)' failed
                                             
Thread 1 "gnome-shell" received signal SIGTRAP, Trace/breakpoint trap.                                                                
_g_log_abort (breakpoint=breakpoint@entry=1) at ../../../../glib/gmessages.c:554
warning: Source file is more recent than executable.
554     }                                        
(gdb) bt
#0  0x00007ffff705c041 in _g_log_abort (breakpoint=breakpoint@entry=1) at ../../../../glib/gmessages.c:554
#1  0x00007ffff705d33d in g_logv (log_domain=0x7ffff709f80e "GLib", log_level=G_LOG_LEVEL_CRITICAL, format=<optimized out>, args=args@entry=0x7fffffffdcd0) at ../../../../glib/gmessages.c:1362
#2  0x00007ffff705d49f in g_log (log_domain=<optimized out>, log_level=<optimized out>, format=<optimized out>) at ../../../../glib/gmessages.c:1403
#3  0x00007fffa75f458e in serialize_attached_effects_to_variant (effects_for_events=0x5555585b7a40) at ../animations-dbus/animations-dbus-server-surface.c:540
#4  0x00007fffa75f47c0 in animations_dbus_server_surface_get_property (object=0x5555585cbb70 [AnimationsDbusServerSurface], prop_id=6, value=0x7fffffffe060, pspec=0x555557389840 [GParamVariant]) at ../animations-dbus/animations-dbus-server-surface.c:592
#5  0x00007ffff733ab51 in object_get_property (value=0x7fffffffe060, pspec=0x555557389840 [GParamVariant], object=0x5555585cbb70 [AnimationsDbusServerSurface]) at ../../../../gobject/gobject.c:1394
#6  0x00007ffff733ab51 in g_object_get_property (object=0x5555585cbb70 [AnimationsDbusServerSurface], property_name=<optimized out>, value=0x7fffffffe060) at ../../../../gobject/gobject.c:2591
```

If I change the `GVariantType` passed to `g_variant_builder_open` to just `"s"`, I get a different assertion:

```
(gnome-shell:65183): GLib-CRITICAL **: g_variant_builder_init: assertion 'g_variant_type_is_container (type)' failed                                         
                                                                                                                                                              
Thread 1 "gnome-shell" received signal SIGTRAP, Trace/breakpoint trap.                                                                                                                                                                                                         
_g_log_abort (breakpoint=breakpoint@entry=1) at ../../../../glib/gmessages.c:554                                                                                                                     
554     ../../../../glib/gmessages.c: No such file or directory.                                                                                                                                
(gdb) bt                                                                                                                                                                                                                                                                        
#0  0x00007ffff705c041 in _g_log_abort (breakpoint=breakpoint@entry=1) at ../../../../glib/gmessages.c:554                                                       
#1  0x00007ffff705d33d in g_logv (log_domain=0x7ffff709f80e "GLib", log_level=G_LOG_LEVEL_CRITICAL, format=<optimized out>, args=args@entry=0x7fffffffdcb0) at ../../../../glib/gmessages.c:1362
#2  0x00007ffff705d49f in g_log (log_domain=<optimized out>, log_level=log_level@entry=G_LOG_LEVEL_CRITICAL, format=format@entry=0x7ffff70a94a4 "%s: assertion '%s' failed") at ../../../../glib/gmessages.c:1403
#3  0x00007ffff705dbb9 in g_return_if_fail_warning (log_domain=<optimized out>, pretty_function=<optimized out>, expression=<optimized out>) at ../../../../glib/gmessages.c:2755                                                                                               
#4  0x00007ffff708fb6e in g_variant_builder_open (builder=0x7fffffffddc0, type=0x7fffa75f7141) at ../../../../glib/gvariant.c:3572                                              
#5  0x00007fffa75f45c4 in serialize_attached_effects_to_variant (effects_for_events=0x5555585e1b00) at ../animations-dbus/animations-dbus-server-surface.c:540
```

It seems that the variant builder wants the opened type to be a subtype of the builder type, but must itself also be a container. If that were the case, how would building a GStrv work?